gilbreth_gazebo/src/plugins/ObjectDisposalPlugin.cc

```diff
@@ -60,6 +68,71 @@ void ObjectDisposalPlugin::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
   }
 
   this->disposalPose = _sdf->Get<math::Pose>("disposal_pose");
+
+  // connecting to ROS
+  GZ_ASSERT(ros::isInitialized(),"ROS is not initialized, can't subscribe to delete model server");
+
+  // setting up ROS connections
+  nh_.reset(new ros::NodeHandle());
+  delete_model_client_ = nh_->serviceClient<gazebo_msgs::DeleteModel>(DELETE_MODEL_SERVICE);
+  set_state_client_ = nh_->serviceClient<gazebo_msgs::SetModelState>(SET_MODEL_STATE_SERVICE);
+  GZ_ASSERT(delete_model_client_.waitForExistence(ros::Duration(WAIT_SERVICE_PERIOD)),
+            boost::str(boost::format("The service '%1%' was not found") % DELETE_MODEL_SERVICE).c_str());
+
+  // setting up info
+  world_frame_id_ = DEFAULT_WORLD_FRAME_ID;
+  gzmsg << boost::str(boost::format("World Frame ID '%1%'") % world_frame_id_)<< std::endl;
+
+  // ROS Thread setup
+  ros_queue_thread_ = std::thread(std::bind(&ObjectDisposalPlugin::processROSQueue,this));
+
+  gzmsg << boost::str(boost::format("Connected to service '%1%'") % DELETE_MODEL_SERVICE)<< std::endl;
+
+
+}
+
+void ObjectDisposalPlugin::processROSQueue()
+{
+  ros::Duration loop_pause(ROS_QUEUE_TIMEOUT);
+  while(nh_->ok())
+  {
+    deleteQueuedObjects();
+    loop_pause.sleep();
+  }
+}
+
+void ObjectDisposalPlugin::deleteQueuedObjects()
+{
+  gazebo_msgs::DeleteModel delete_model_srv;
+  gazebo_msgs::SetModelState set_state_srv;
+  while(!delete_model_queue_.empty())
+  {
+    std::string model_name = delete_model_queue_.front();
+
+    // moving model out of the way
+    gazebo_msgs::SetModelStateRequest& set_st_req = set_state_srv.request;
+    set_st_req.model_state.model_name = model_name;
+    Eigen::Affine3d pose = Eigen::Affine3d::Identity() * Eigen::Translation3d(DISPOSE_LOCATION);
+    tf::poseEigenToMsg(pose,set_st_req.model_state.pose);
+    set_st_req.model_state.reference_frame = world_frame_id_;
+
+    if(!set_state_client_.call(set_state_srv))
+    {
+      gzwarn<<boost::str(boost::format("Failed to move model '%1%' to discard location")% model_name)<<std::endl;
+    }
+
+    delete_model_srv.request.model_name = model_name;
+    if(delete_model_client_.call(delete_model_srv) && delete_model_srv.response.success)
+    {
+      gzdbg <<"Model '"<< delete_model_srv.request.model_name <<"' deleted"<<std::endl;
+    }
+    else
+    {
+      gzerr <<"Model deletion failed for '"<< delete_model_srv.request.model_name <<"'" <<std::endl;
+    }
+
+    delete_model_queue_.pop();
+  }
 }
```

Comment:
Will this plugin dispose objects in the bin?

Reply:
The service call in [line 125](https://github.com/swri-robotics/gilbreth/pull/42/files/430f8c64e0a6c58e11936de4261b75f8f984c82b..83e49a839def9cc087fb85e05701b6cea331de67#diff-b00a4c7b5be891f207e016130e40ae93R125) is what actually deletes the object, the `pop()` method just removes the object's name from  a concurrent queue so that a second attempt to remove it is not made.